From 14f2577f834fe3cde5e6ee4a90c97a468940502f Mon Sep 17 00:00:00 2001
From: Jan Moeller <Jan.Moeller@governikus.de>
Date: Tue, 26 May 2026 14:50:45 +0200
Subject: a11y: Introduce QAccessibleViewportInterface

This interface allows the platform specific a11y bridges
to supply the screen reader with sufficient information
about the current portion of the viewport.

For something like QML Flickable this might be the current
position within the entire content width/height. Android
TalkBack emits "ding ding" sounds depending on the viewport
within the underlying content.

For ListViews this may be used to inform a11y users about
the current of elements within the viewport and the total
amount of items.

Task-number: QTBUG-139833
Change-Id: Id16e74627663543cdd2918afcd8a24adce6470ff
Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
(cherry-picked from commit 89ea7832e7d5bbed47271bea3df14e37531c9a4b)
---
 src/gui/accessible/qaccessible.cpp    | 68 +++++++++++++++++++++++++--
 src/gui/accessible/qaccessible.h      | 16 +++++++
 src/gui/accessible/qaccessible_base.h |  1 +
 3 files changed, 82 insertions(+), 3 deletions(-)

diff --git x/qtbase/src/gui/accessible/qaccessible.cpp y/qtbase/src/gui/accessible/qaccessible.cpp
index 9886f1420bdd9a3a0cec5c859e8aa6181d955c69..67ad4b9c73cef5f9a97a2b66497e5a593bf6f2e3 100644
--- x/qtbase/src/gui/accessible/qaccessible.cpp
+++ y/qtbase/src/gui/accessible/qaccessible.cpp
@@ -511,8 +511,9 @@ Q_STATIC_LOGGING_CATEGORY(lcAccessibilityCore, "qt.accessibility.core");
     \value HyperlinkInterface       For hyperlink nodes (usually embedded as children of text nodes)
     \value [since 6.5] SelectionInterface For non-text objects that support selection of child objects.
     \value [since 6.8] AttributesInterface For objects that support object-specific attributes.
+    \value [since 6.12] ViewportInterface For objects that allow 2-dimensional or indexed scrolling.
 
-    \sa QAccessibleInterface::interface_cast(), QAccessibleTextInterface, QAccessibleValueInterface, QAccessibleActionInterface, QAccessibleTableInterface, QAccessibleTableCellInterface, QAccessibleSelectionInterface, QAccessibleAttributesInterface
+    \sa QAccessibleInterface::interface_cast(), QAccessibleTextInterface, QAccessibleValueInterface, QAccessibleActionInterface, QAccessibleTableInterface, QAccessibleTableCellInterface, QAccessibleSelectionInterface, QAccessibleAttributesInterface, QAccessibleViewportInterface
 */
 
 #if QT_CONFIG(accessibility)
@@ -1127,7 +1128,9 @@ std::pair< int, int > QAccessible::qAccessibleTextBoundaryHelper(const QTextCurs
 
     Lists, tables and trees should implement QAccessibleTableInterface.
 
-    \sa QAccessible, QAccessibleActionInterface, QAccessibleTextInterface, QAccessibleValueInterface, QAccessibleTableInterface
+    Scrollable views should implement QAccessibleViewportInterface.
+
+    \sa QAccessible, QAccessibleActionInterface, QAccessibleTextInterface, QAccessibleValueInterface, QAccessibleTableInterface, QAccessibleViewportInterface
 */
 
 /*!
@@ -1390,6 +1393,11 @@ QColor QAccessibleInterface::backgroundColor() const
     \since 6.5
 */
 
+/*!
+    \fn QAccessibleViewportInterface *QAccessibleInterface::viewportInterface()
+    \since 6.12
+*/
+
 /*!
     \class QAccessibleEvent
     \ingroup accessibility
@@ -1965,7 +1973,8 @@ void QAccessibleInterface::virtual_hook(int /*id*/, void * /*data*/)
 
     \sa QAccessible::InterfaceType, QAccessibleTextInterface,
     QAccessibleValueInterface, QAccessibleActionInterface,
-    QAccessibleTableInterface, QAccessibleTableCellInterface
+    QAccessibleTableInterface, QAccessibleTableCellInterface,
+    QAccessibleViewportInterface
 */
 
 /*! \internal */
@@ -3341,6 +3350,59 @@ QAccessibleHyperlinkInterface::~QAccessibleHyperlinkInterface()
 
 }
 
+/*!
+    \since 6.12
+    \class QAccessibleViewportInterface
+    \inmodule QtGui
+    \ingroup accessibility
+    \preliminary
+
+    \brief The QAccessibleViewportInterface class implements support for viewports.
+
+    This interface should be implemented by accessible objects that allow an 2-dimensional
+    or indexed viewports (e.g. by scrolling). For example lists or flickable views.
+*/
+
+/*!
+    Destroys the QAccessibleViewportInterface.
+
+*/
+QAccessibleViewportInterface::~QAccessibleViewportInterface()
+{
+}
+
+/*!
+    \fn QSizeF QAccessibleViewportInterface::contentSize() const
+
+    Returns the total size of the content in absolute values. For non-indexed viewports this will return
+    the absolute size of the underlying content, for indexed viewports the count of rows and columns is
+    returned.
+*/
+
+/*!
+    \fn QPointF QAccessibleViewportInterface::position() const
+
+    Returns the current position within the viewport in normalized values.
+*/
+
+/*!
+    \fn QSizeF QAccessibleViewportInterface::viewportSize() const
+
+    Returns the currently visible portion of the content in normalized values.
+*/
+
+/*!
+    \fn bool QAccessibleViewportInterface::isIndexed() const
+
+    Returns true if the underlying content is indexed (e.g. a list of items).
+*/
+
+/*!
+    \fn void QAccessibleViewportInterface::setPosition(const QPointF &position)
+
+    Sets the viewport's position to the given \a position defined in normalized values.
+*/
+
 #endif // QT_CONFIG(accessibility)
 
 QT_END_NAMESPACE
diff --git x/qtbase/src/gui/accessible/qaccessible.h y/qtbase/src/gui/accessible/qaccessible.h
index 9aa223ed8b3a969a7ceb00adbb3386022fb11897..c15f4af70dbb22d07ec31df39e841d3851a7c537 100644
--- x/qtbase/src/gui/accessible/qaccessible.h
+++ y/qtbase/src/gui/accessible/qaccessible.h
@@ -45,6 +45,7 @@ class QAccessibleHyperlinkInterface;
 class QAccessibleSelectionInterface;
 class QAccessibleAttributesInterface;
 class QAccessibleTableModelChangeEvent;
+class QAccessibleViewportInterface;
 
 class Q_GUI_EXPORT QAccessibleInterface
 {
@@ -110,6 +111,9 @@ public:
     inline QAccessibleAttributesInterface *attributesInterface()
     { return reinterpret_cast<QAccessibleAttributesInterface *>(interface_cast(QAccessible::AttributesInterface)); }
 
+    inline QAccessibleViewportInterface *viewportInterface()
+    { return reinterpret_cast<QAccessibleViewportInterface *>(interface_cast(QAccessible::ViewportInterface)); }
+
     virtual void virtual_hook(int id, void *data);
 
     virtual void *interface_cast(QAccessible::InterfaceType)
@@ -297,6 +301,18 @@ public:
     virtual QVariant attributeValue(QAccessible::Attribute key) const = 0;
 };
 
+class Q_GUI_EXPORT QAccessibleViewportInterface
+{
+public:
+    virtual ~QAccessibleViewportInterface();
+
+    virtual QSizeF contentSize() const = 0;
+    virtual QPointF position() const = 0;
+    virtual QSizeF viewportSize() const = 0;
+    virtual bool isIndexed() const = 0;
+    virtual void setPosition(const QPointF &position) = 0;
+};
+
 
 class Q_GUI_EXPORT QAccessibleEvent
 {
diff --git x/qtbase/src/gui/accessible/qaccessible_base.h y/qtbase/src/gui/accessible/qaccessible_base.h
index 177fc9ad5d693d6bc98c2ca1e2f43fccb522a232..ec88ecf9ad446026b4835b2ddb313a364b3e5848 100644
--- x/qtbase/src/gui/accessible/qaccessible_base.h
+++ y/qtbase/src/gui/accessible/qaccessible_base.h
@@ -368,6 +368,7 @@ public:
         HyperlinkInterface,
         SelectionInterface,
         AttributesInterface,
+        ViewportInterface,
     };
 
     enum TextBoundaryType {
